From e7f56826c893dd67687a77011adbf5d7d2757770 Mon Sep 17 00:00:00 2001
From: lagosproject <lakescorporation@gmail.com>
Date: Wed, 16 Sep 2026 12:28:19 +0200
Subject: [PATCH] Add Tanix TX6s (H616 + AXP305) device tree and defconfig

Tanix TX6s board with Allwinner H616, DDR3 @ 648 MHz, AXP305 PMIC via RSB.
Supports eMMC and MicroSD. Verified on physical hardware.

Signed-off-by: lagosproject <lakescorporation@gmail.com>
---
 configs/tanix_tx6s_defconfig                  |  38 +++
 .../allwinner/sun50i-h616-tanix-tx6s.dts      | 254 ++++++++++++++++++
 2 files changed, 292 insertions(+)
 create mode 100644 configs/tanix_tx6s_defconfig
 create mode 100644 dts/upstream/src/arm64/allwinner/sun50i-h616-tanix-tx6s.dts

diff --git a/configs/tanix_tx6s_defconfig b/configs/tanix_tx6s_defconfig
new file mode 100644
index 0000000..e14199d
--- /dev/null
+++ b/configs/tanix_tx6s_defconfig
@@ -0,0 +1,38 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="allwinner/sun50i-h616-tanix-tx6s"
+CONFIG_SPL=y
+CONFIG_SUNXI_DRAM_H616_DDR3_1333=y
+CONFIG_DRAM_CLK=648
+CONFIG_DRAM_ODT_EN=y
+CONFIG_DRAM_SUNXI_UNKNOWN_FEATURE=y
+CONFIG_DRAM_SUNXI_BIT_DELAY_COMPENSATION=y
+CONFIG_DRAM_SUNXI_READ_CALIBRATION=y
+CONFIG_DRAM_SUNXI_DX_ODT=0x03030303
+CONFIG_DRAM_SUNXI_DX_DRI=0x0e0e0e0e
+CONFIG_DRAM_SUNXI_CA_DRI=0x00001c12
+CONFIG_DRAM_SUNXI_ODT_EN=0x00000001
+CONFIG_DRAM_SUNXI_TPR0=0xc0000c05
+CONFIG_DRAM_SUNXI_TPR2=0x00000000
+CONFIG_DRAM_SUNXI_TPR6=0x33808080
+CONFIG_DRAM_SUNXI_TPR10=0x2f0007
+CONFIG_DRAM_SUNXI_TPR11=0xffffdddd
+CONFIG_DRAM_SUNXI_TPR12=0xfedf7557
+CONFIG_MACH_SUN50I_H616=y
+CONFIG_AXP305_POWER=y
+CONFIG_R_I2C_ENABLE=y
+CONFIG_SPL_I2C=y
+CONFIG_SPL_I2C_SUPPORT=y
+CONFIG_SPL_SYS_I2C_LEGACY=y
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_SYS_I2C_SLAVE=0x7f
+CONFIG_SYS_I2C_SPEED=100000
+CONFIG_PHY_REALTEK=y
+CONFIG_SUN8I_EMAC=y
+CONFIG_I2C3_ENABLE=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_MUSB_GADGET=y
+CONFIG_SUPPORT_EMMC_BOOT=y
+CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR=0x40
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
diff --git a/dts/upstream/src/arm64/allwinner/sun50i-h616-tanix-tx6s.dts b/dts/upstream/src/arm64/allwinner/sun50i-h616-tanix-tx6s.dts
new file mode 100644
index 0000000..da0fb5d
--- /dev/null
+++ b/dts/upstream/src/arm64/allwinner/sun50i-h616-tanix-tx6s.dts
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: (GPL-2.0+ or MIT)
+/*
+ * Copyright (C) 2021 Arm Ltd.
+ */
+
+/dts-v1/;
+
+#include "sun50i-h616.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "Tanix TX6s";
+	compatible = "tanix,tx6s", "allwinner,sun50i-h616";
+
+	aliases {
+		mmc0 = &mmc0;
+		mmc1 = &mmc1;
+		mmc2 = &mmc2;
+		ethernet0 = &emac1;
+		serial0 = &uart0;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	reg_vcc5v: vcc5v {
+		/* board wide 5V supply directly from the DC input */
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-5v";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-always-on;
+	};
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ir {
+	status = "okay";
+};
+
+&pio {
+	rmii_pins: rmii-pins {
+		pins = "PA0", "PA1", "PA2", "PA3", "PA4",
+		       "PA5", "PA6", "PA7", "PA8", "PA9";
+		function = "emac1";
+		drive-strength = <40>;
+	};
+};
+
+/ {
+	soc {
+		emac1: ethernet@5030000 {
+			compatible = "allwinner,sun50i-h616-internal-emac";
+			syscon = <&syscon 1>;
+			reg = <0x05030000 0x10000>;
+			interrupts = <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "macirq";
+			resets = <&ccu RST_BUS_EMAC1>;
+			reset-names = "stmmaceth";
+			clocks = <&ccu CLK_BUS_EMAC1>, <&ccu CLK_EMAC_25M>;
+			clock-names = "stmmaceth", "emac-25m";
+			pinctrl-names = "default";
+			pinctrl-0 = <&rmii_pins>;
+			phy-mode = "rmii";
+			phy-handle = <&rmii_phy>;
+			phy-supply = <&reg_aldo1>;
+			allwinner,rx-delay-ps = <3100>;
+			allwinner,tx-delay-ps = <700>;
+			status = "okay";
+
+			mdio1: mdio {
+				compatible = "snps,dwmac-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				rmii_phy: ethernet-phy@1 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <1>;
+				};
+			};
+		};
+	};
+};
+
+&mmc0 {
+	vmmc-supply = <&reg_dcdce>;
+	//cd-gpios = <&pio 8 16 GPIO_ACTIVE_LOW>;	/* PI16 */
+	broken-cd;
+	bus-width = <4>;
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&reg_dcdce>;
+	broken-cd;
+	bus-width = <4>;
+	status = "okay";
+};
+
+&mmc2 {
+	vmmc-supply = <&reg_dcdce>;
+	bus-width = <8>;
+	non-removable;
+	cap-mmc-hw-reset;
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
+&r_rsb {
+	status = "okay";
+
+	axp305: pmic@745 {
+		compatible = "x-powers,axp305", "x-powers,axp805",
+			     "x-powers,axp806";
+		interrupt-controller;
+		#interrupt-cells = <1>;
+		reg = <0x745>;
+
+		x-powers,self-working-mode;
+		vina-supply = <&reg_vcc5v>;
+		vinb-supply = <&reg_vcc5v>;
+		vinc-supply = <&reg_vcc5v>;
+		vind-supply = <&reg_vcc5v>;
+		vine-supply = <&reg_vcc5v>;
+		aldoin-supply = <&reg_vcc5v>;
+		bldoin-supply = <&reg_vcc5v>;
+		cldoin-supply = <&reg_vcc5v>;
+
+		regulators {
+			reg_aldo1: aldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc-sys";
+			};
+
+			reg_aldo2: aldo2 {
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc3v3-ext";
+				status = "disabled";
+			};
+
+			reg_aldo3: aldo3 {
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc3v3-ext2";
+				status = "disabled";
+			};
+
+			reg_bldo1: bldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <1800000>;
+				regulator-name = "vcc1v8";
+			};
+
+			reg_bldo2: bldo2 {
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <1800000>;
+				regulator-name = "vcc1v8-2";
+				status = "disabled";
+			};
+
+			bldo3 {
+				/* unused */
+			};
+
+			bldo4 {
+				/* unused */
+			};
+
+			cldo1 {
+				regulator-min-microvolt = <2500000>;
+				regulator-max-microvolt = <2500000>;
+				regulator-name = "vcc2v5";
+			};
+
+			cldo2 {
+				/* unused */
+			};
+
+			cldo3 {
+				/* unused */
+			};
+
+			reg_dcdca: dcdca {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <1080000>;
+				regulator-name = "vdd-cpu";
+			};
+
+			reg_dcdcc: dcdcc {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <1080000>;
+				regulator-name = "vdd-gpu-sys";
+			};
+
+			reg_dcdcd: dcdcd {
+				regulator-always-on;
+				regulator-min-microvolt = <1360000>;
+				regulator-max-microvolt = <1360000>;
+				regulator-name = "vdd-dram";
+			};
+
+			reg_dcdce: dcdce {
+				regulator-boot-on;
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc-eth-mmc";
+			};
+
+			sw {
+				/* unused */
+			};
+		};
+	};
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_ph_pins>;
+	status = "okay";
+};
+
+&usbotg {
+	dr_mode = "host";	/* USB A type receptable */
+	status = "okay";
+};
+
+&usbphy {
+	status = "okay";
+};
+
-- 
2.43.0

